pkg/agent/controller/networkpolicy/fqdn_test.go

```diff
@@ -584,3 +584,112 @@ func TestSyncDirtyRules(t *testing.T) {
+				"192.1.1.2": currentTime.Add(6 * time.Second),
+				"192.1.1.3": currentTime.Add(10 * time.Second),
+			},
+		},
+		{
+			name: "old IP resent in DNS response is retained with an updated TTL fetched from response",
+			existingDNSCache: map[string]dnsMeta{
+				"fqdn-test-pod.lfx.test": {
+					responseIPs: map[string]ipWithTTL{
+						"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(5 * time.Second)},
+					},
+				},
+			},
+			dnsResponseIPs: map[string]ipWithTTL{
+				"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(10 * time.Second)},
+			},
+			expectedIPs: map[string]time.Time{
+				"192.1.1.1": currentTime.Add(10 * time.Second),
+			},
+		},
+
+		{
+			name: "stale IP with expired TTL is evicted",
+			existingDNSCache: map[string]dnsMeta{
+				"fqdn-test-pod.lfx.test": {
+					responseIPs: map[string]ipWithTTL{
+						"192.1.1.1": {ip: net.ParseIP("192.1.1.1"), expirationTime: currentTime.Add(-1 * time.Second)},
+					},
+				},
+			},
+			dnsResponseIPs: map[string]ipWithTTL{
+				"192.1.1.3": {ip: net.ParseIP("192.1.1.3"), expirationTime: currentTime.Add(10 * time.Second)},
+			},
+			expectedIPs: map[string]time.Time{
+				"192.1.1.3": currentTime.Add(10 * time.Second),
+			},
+		},
+		{
+			name: "stale IP with unexpired TTL are retained",
+			existingDNSCache: map[string]dnsMeta{
+				"fqdn-test-pod.lfx.test": {
+					responseIPs: map[string]ipWithTTL{
+						"192.1.1.2": {ip: net.ParseIP("192.1.1.2"), expirationTime: currentTime.Add(5 * time.Second)},
+					},
+				},
+			},
+			dnsResponseIPs: map[string]ipWithTTL{
+				"192.1.1.3": {ip: net.ParseIP("192.1.1.3"), expirationTime: currentTime.Add(10 * time.Second)},
+			},
+			expectedIPs: map[string]time.Time{
+				"192.1.1.2": currentTime.Add(5 * time.Second),
+				"192.1.1.3": currentTime.Add(10 * time.Second),
+			},
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			controller := gomock.NewController(t)
+			f, _ := newMockFQDNController(t, controller, nil)
+			f.dnsEntryCache = tc.existingDNSCache
+
+			f.onDNSResponse("fqdn-test-pod.lfx.test", tc.dnsResponseIPs, nil)
+
+			cachedDnsMetaData := f.dnsEntryCache["fqdn-test-pod.lfx.test"]
+			assert.Equal(t, len(cachedDnsMetaData.responseIPs), len(tc.expectedIPs), "Expected %d IPs in cache, got %d", len(tc.expectedIPs), len(cachedDnsMetaData.responseIPs))
+
+			for ipStr, expectedTTL := range tc.expectedIPs {
+				cachedIPMeta, exists := cachedDnsMetaData.responseIPs[ipStr]
+				if !exists {
+					t.Errorf("Expected %s to be found in dns cache.", ipStr)
+					continue
+				}
+
+				if cachedIPMeta.expirationTime.Before(time.Now()) {
+					t.Errorf("Expected %s to be found with a valid TTL,but it has expired.", ipStr)
+				}
+
+				if !cachedIPMeta.expirationTime.Equal(expectedTTL) {
+					t.Errorf("Expected TTL for %s to be %v, got %v", ipStr, expectedTTL, cachedIPMeta.expirationTime)
+				}
+			}
```

Comment:
we should check if the fqdn has been queued  to `dnsQueryQueue` (when it is expected)